### PR TITLE
Homework4 Submission

### PR DESCRIPTION
**Pull request by dataovoxo**

### CODE REVIEW

**Comment by julie-scherer** [comment]
### Feedback:
### Coverage of Homework Prompt
The student's solution covers the homework prompt well. They have converted two queries from TrinoSQL to SparkSQL and created new PySpark jobs for these queries. The student has also created unit tests for these jobs, which use fake input and expected output data. 

### Logical Structure
The student's solution follows a logical structure. The PySpark jobs are well-structured and the SQL queries are correctly converted from TrinoSQL to SparkSQL. The unit tests are also logically structured, with clear setup of test data, execution of the job functions, and comparison of the actual and expected results.

### Readability
The student's code is highly readable. They have used meaningful variable names and provided clear comments explaining the purpose of each function and the logic behind the SQL queries. The code is also well-formatted, making it easy to follow the flow of logic.

### Robustness and Error Handling
The student's code does not include explicit error handling. However, the use of PySpark's built-in functions and SQL queries should ensure that the code is robust and can handle a variety of input data. The unit tests also help to ensure that the code functions as expected. It would be beneficial for the student to include some error handling or validation checks in their code to handle potential issues with the input data or the execution of the SQL queries.

### Modularity and Reusability
The student's code is modular and reusable. The PySpark jobs are defined as separate functions, which can be easily reused in other contexts. The SQL queries are defined as separate functions, which return the query as a string. This makes it easy to reuse these queries in different contexts or to modify them as needed. The unit tests are also modular, with separate tests for each job function.

### Suggestions for Improvement
Overall, the student's solution is well-implemented and covers all aspects of the homework prompt. However, there are a few areas where the student could improve their solution:

- **Error Handling**: The student could add some error handling or validation checks to their code to handle potential issues with the input data or the execution of the SQL queries.
- **Code Comments**: While the student has provided clear comments explaining the purpose of each function, they could also add comments explaining the logic behind the SQL queries. This would make it easier for others to understand how the queries work and how they transform the data.
- **Unit Tests**: The student could add more unit tests to cover different edge cases or potential issues with the input data. This would help to ensure

### Grade:
### Query Conversion

**Grade: Proficient**

The student has successfully converted the TrinoSQL queries to SparkSQL. The queries are correctly structured and maintain the original functionalities. The student has demonstrated a strong understanding of both TrinoSQL and SparkSQL query languages.

### PySpark Jobs

**Grade: Proficient**

The student has created two PySpark jobs that are well-structured and error-free. The jobs demonstrate a robust understanding of Spark capabilities. The jobs are designed to be reusable and maintainable, with the output table name passed as a parameter.

### Unit Tests

**Grade: Proficient**

The student has created comprehensive unit tests for both jobs. The tests cover the main functionalities and ensure the integrity of the code. The tests use named tuples to create the input and expected data, which makes the tests easy to understand and maintain. The use of the `chispa` library for DataFrame comparison is a good choice as it simplifies the

File: src/jobs/job_2.py [hunk 1]
```diff
@@ -4,7 +4,33 @@
 
 def query_2(output_table_name: str) -> str:
     query = f"""
-    <YOUR QUERY HERE>
+    WITH
+    previous_date_records AS (
+        SELECT *
+        FROM {output_table_name}
+        WHERE date = DATE('2023-01-06')
```
Comment (HjersmanJ) [inline src/jobs/job_2.py:11]: Date would be best as a variable.

File: src/jobs/job_1.py [hunk 2]
```diff
@@ -4,7 +4,48 @@
 
 def query_1(output_table_name: str) -> str:
     query = f"""
-    <YOUR QUERY HERE>
+    WITH 
+        previous_year AS (
+            SELECT * 
+            FROM {output_table_name}
+            WHERE current_year = 2013
```
Comment (HjersmanJ) [inline src/jobs/job_1.py:11]: Year being a variable would be ideal.

File: src/unit_tests/test_spark_jobs.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    # WebEvents follows the schema needed for web_events
+    # Devices follows the schema needed for devices
+    # UserDevices defines the schema for the output table and existing data (user_cummulative_devices)
+    WebEvents = namedtuple("WebEvents", "user_id device_id referrer	host url event_time") 
+    Devices = namedtuple("Devices", "browser_type device_id")
+    UserDevices = namedtuple("UserDevices", "user_id browser_type dates_active date")
+
+    # Define the input data for testing. Input define for web_events and devices
+    input_events_data = [
+        WebEvents(-9999999999, 12345678, 'www.google.com', 'www.eczachly.com', '/', datetime(2023, 1, 7, 10, 30, tzinfo=pytz.UTC))
```
Comment (HjersmanJ) [inline src/unit_tests/test_spark_jobs.py:149]: Unless clearly stated, I'd avoid putting negative inputs for user_id

**Review by HjersmanJ — COMMENTED** [review]
You completed the requirements for the assignment, but this would fit the purpose more with a few changes. Make items variables that can allow them to be reusable. Changing a year on an incremental refresh query is an example. Also try to keep data consistent with what people expect. I don't think I've ever seen negative numbers for user_id. However, I've seen people use them as a placeholder or an error log. If that is the case, it should be stated as such, though that's not usually good practice. 

Good job on the assignment!
